command/apply.go

```diff
@@ -144,103 +206,75 @@ func (c *ApplyCommand) Run(args []string) int {
+	opReq.PlanRefresh = args.Refresh
+	opReq.Targets = args.Targets
 	opReq.Type = backend.OperationTypeApply
-	opReq.View = views.NewOperation(arguments.ViewHuman, c.RunningInAutomation, c.View)
-
-	opReq.ConfigLoader, err = c.initConfigLoader()
-	if err != nil {
-		c.showDiagnostics(err)
-		return 1
+	opReq.View = view.Operation()
+	// FIXME: this shim is needed until the remote backend is migrated to views
+	opReq.ShowDiagnostics = func(vals ...interface{}) {
+		var diags tfdiags.Diagnostics
+		diags = diags.Append(vals...)
+		view.Diagnostics(diags)
 	}
 
-	{
-		var moreDiags tfdiags.Diagnostics
-		opReq.Variables, moreDiags = c.collectVariableValues()
-		diags = diags.Append(moreDiags)
-		if moreDiags.HasErrors() {
-			c.showDiagnostics(diags)
-			return 1
-		}
-	}
-
-	op, err := c.RunOperation(be, opReq)
+	var err error
+	opReq.ConfigLoader, err = c.initConfigLoader()
 	if err != nil {
-		c.showDiagnostics(err)
-		return 1
-	}
-
-	if op.Result != backend.OperationSuccess {
-		return op.Result.ExitStatus()
+		diags = diags.Append(fmt.Errorf("Failed to initialize config loader: %s", err))
+		return nil, diags
 	}
 
-	// Show the count results from the operation
-	if c.Destroy {
-		c.Ui.Output(c.Colorize().Color(fmt.Sprintf(
-			"[reset][bold][green]\n"+
-				"Destroy complete! Resources: %d destroyed.",
-			countHook.Removed)))
-	} else {
-		c.Ui.Output(c.Colorize().Color(fmt.Sprintf(
-			"[reset][bold][green]\n"+
-				"Apply complete! Resources: %d added, %d changed, %d destroyed.",
-			countHook.Added,
-			countHook.Changed,
-			countHook.Removed)))
-	}
-
-	// only show the state file help message if the state is local.
-	if (countHook.Added > 0 || countHook.Changed > 0) && c.Meta.stateOutPath != "" {
-		c.Ui.Output(c.Colorize().Color(fmt.Sprintf(
-			"[reset]\n"+
-				"The state of your infrastructure has been saved to the path\n"+
-				"below. This state is required to modify and destroy your\n"+
-				"infrastructure, so keep it safe. To inspect the complete state\n"+
-				"use the `terraform show` command.\n\n"+
-				"State path: %s",
-			c.Meta.stateOutPath)))
-	}
+	return opReq, diags
+}
 
-	if !c.Destroy && op.State != nil {
-		outputValues := op.State.RootModule().OutputValues
-		if len(outputValues) > 0 {
-			c.Ui.Output(c.Colorize().Color("[reset][bold][green]\nOutputs:\n\n"))
-			view := views.NewOutput(arguments.ViewHuman, c.View)
-			view.Output("", outputValues)
-		}
+func (c *ApplyCommand) GatherVariables(opReq *backend.Operation, args *arguments.Vars) tfdiags.Diagnostics {
+	var diags tfdiags.Diagnostics
+
+	// FIXME the arguments package currently trivially gathers variable related
+	// arguments in a heterogenous slice, in order to minimize the number of
+	// code paths gathering variables during the transition to this structure.
+	// Once all commands that gather variables have been converted to this
+	// structure, we could move the variable gathering code to the arguments
+	// package directly, removing this shim layer.
```

Comment:
FIXME! The comment sums it up. I think the ideal outcome of this change would be for the arguments package to gather all variables from flags and the environment, and return `(map[string]backend.UnparsedVariableValue, tfdiags.Diagnostics)`. This would allow this awkward type conversion and keep the flag package hacking isolated to the arguments package.